File: src/main/java/toast/client/modules/render/BlockESP.java
package toast.client.modules.render;

import com.google.common.eventbus.Subscribe;
import net.minecraft.block.Block;
import net.minecraft.network.packet.s2c.play.ChunkDataS2CPacket;
import net.minecraft.util.math.BlockPos;
import toast.client.events.network.EventPacketReceived;
import toast.client.modules.Module;
import toast.client.utils.WorldUtil;

import java.util.List;

public class BlockESP extends Module {
    public static List<Block> matches;
    public static List<BlockPos> found;

    private static List<Block> checkedChunks;

    public BlockESP() {
        super("BlockESP", "Highlights blocks in the world.", Category.RENDER, -1);
    }

    @Subscribe
    public void onPacketReceived(EventPacketReceived event) {
        if (mc.world == null) return;
        if (event.getPacket() instanceof ChunkDataS2CPacket) {
            ChunkDataS2CPacket packet = (ChunkDataS2CPacket) event.getPacket();
            try {
                WorldUtil.searchChunk(mc.world, packet.getX() / 16, packet.getZ() / 16, matches).keySet().addAll(found);
            } catch (InterruptedException ignored) {
            }
        }
    }
}

File: src/main/java/toast/client/modules/render/BlockESP.kt
package toast.client.modules.render

import com.google.common.eventbus.Subscribe
import net.minecraft.block.Block
import net.minecraft.network.packet.s2c.play.BlockUpdateS2CPacket
import net.minecraft.util.math.BlockPos
import toast.client.events.network.EventPacketReceived
import toast.client.modules.Module

/**
 * Searches for and highlights blocks in the world
 */
class BlockESP : Module("BlockESP", "Highlights blocks in the world.", Category.RENDER, -1) {
    override fun onEnable() {
        disable()
    }

    @Subscribe
    fun onPacketReceived(event: EventPacketReceived) {
        if (mc.world == null) return
        if (event.getPacket() is BlockUpdateS2CPacket) {
            val packet : BlockUpdateS2CPacket = event.getPacket() as BlockUpdateS2CPacket
            if (matches.contains(packet.state.block)) {
                found.add(packet.pos)
            }
        }
    }

    companion object {
        /**
         * List of blocks to search for
         */
        var matches: List<Block> = ArrayList()

        /**
         * List of coordinates of the matching blocks
         */
        var found: MutableList<BlockPos> = mutableListOf()
    }
}